File: backend/src/main/scala/slaydemo/backend/battle/objects/BattleCommandSkillIntents.scala
package slaydemo.backend.battle.objects

final case class BattleCommandSkillIntents(values: Vector[SkillKind]) {
  def nonEmpty: Boolean =
    values.nonEmpty
}

object BattleCommandSkillIntents {
  val empty: BattleCommandSkillIntents =
    BattleCommandSkillIntents(Vector.empty)

  def fromLegacyFlags(
    castDash: Boolean,
    castBlink: Boolean,
    castFreeze: Boolean
  ): BattleCommandSkillIntents =
    BattleCommandSkillIntents(
      Vector(
        Option.when(castBlink)(SkillKind.Blink),
        Option.when(castDash)(SkillKind.Dash),
        Option.when(castFreeze)(SkillKind.Freeze)
      ).flatten
    )
}

File: frontend/src/objects/battle/microservices/abilities/objects/skill/BattleCommandSkillIntents.ts
import type { SkillKind } from "./SkillKind";

export interface BattleCommandSkillIntents {
  values: SkillKind[];
}

export const emptyBattleCommandSkillIntents: BattleCommandSkillIntents = {
  values: []
};

export function battleCommandSkillIntentsFromLegacyFlags(input: {
  castDash: boolean;
  castBlink: boolean;
  castFreeze: boolean;
  castCritical: boolean;
}): BattleCommandSkillIntents {
  return {
    values: [
      input.castBlink ? "Blink" : null,
      input.castDash ? "Dash" : null,
      input.castFreeze ? "Freeze" : null,
      input.castCritical ? "Critical" : null
    ].filter((value): value is SkillKind => value !== null)
  };
}
